src/Routers/FilesRouter.js

```diff
@@ -87,41 +112,121 @@ export class FilesRouter {
-        next(
-          new Parse.Error(
-            Parse.Error.FILE_SAVE_ERROR,
-            `Could not store file: ${filename}.`
-          )
+    const base64 = req.body.toString('base64');
+    const file = new Parse.File(filename, { base64 }, contentType);
+    const { metadata = {}, tags = {} } = req.fileData || {};
+    file.setTags(tags);
+    file.setMetadata(metadata);
+    const fileSize = Buffer.byteLength(req.body);
+    const fileObject = { file, fileSize };
+    try {
+      // run beforeSaveFile trigger
+      const triggerResult = await triggers.maybeRunFileTrigger(
+        triggers.Types.beforeSaveFile,
+        fileObject,
+        config,
+        req.auth
+      )
+      let saveResult;
+      // if a new ParseFile is returned check if it's an already saved file
+      if (triggerResult instanceof Parse.File) {
+        fileObject.file = triggerResult;
+        if (triggerResult.url()) {
+          // set fileSize to null because we wont know how big it is here
+          fileObject.fileSize = null;
+          saveResult = {
+            url: triggerResult.url(),
+            name: triggerResult._name,
+          };
+        }
+      }
+      // if the file returned by the trigger has already been saved skip saving anything
+      if (!saveResult) {
+        // if the ParseFile returned is type uri, download the file before saving it
+        await addFileDataIfNeeded(fileObject.file);
+        // update fileSize
+        const bufferData = Buffer.from(fileObject.file._data, 'base64');
+        fileObject.fileSize = Buffer.byteLength(bufferData);
+        // save file
+        const createFileResult = await filesController.createFile(
+          config,
+          fileObject.file._name,
+          bufferData,
+          fileObject.file._source.type,
+          {
+            tags: fileObject.file._tags,
+            metadata: fileObject.file._metadata,
+          }
         );
-      });
+        // update file with new data
+        fileObject.file._name = createFileResult.name;
+        fileObject.file._url = createFileResult.url;
+        fileObject.file._requestTask = null;
+        fileObject.file._previousSave = Promise.resolve(fileObject.file);
+        saveResult = {
+          url: createFileResult.url,
+          name: createFileResult.name,
+        };
+      }
+      // run afterSaveFile trigger
+      await triggers.maybeRunFileTrigger(
+        triggers.Types.afterSaveFile,
+        fileObject,
+        config,
+        req.auth
+      );
+      res.status(201);
+      res.set('Location', saveResult.url);
+      res.json(saveResult);
+
+    } catch (e) {
+      // TODO: Should the error message from a throw in beforeSaveFile hook be used here (instead of `Could not store file: ${filename}`)?
+      logger.error('Error creating a file: ', e);
+      next(
+        new Parse.Error(
+          Parse.Error.FILE_SAVE_ERROR,
+          `Could not store file: ${fileObject.file._name}.`
```

Comment:
@davimacedo Do you see any harm in changing this line to:

`e.message ? e.message : `Could not store file: ${fileObject.file._name}.``

If I throw an error in the beforeSaveFile trigger, I have no way of getting the message down to the client

Reply:
I don't see any problem.

Reply:
This update has been made